### PR TITLE
Move legacy ciphers into the legacy provider

### PR DESCRIPTION
**Pull request by slontis**
DES, idea, seed, rc2, rc4, rc5, cast and blowfish have been moved out of the default provider.

Code shared between desx and tdes has been moved into a seperate file (cipher_tdes_common.c).

3 test recipes failed due to using app/openssl calls that used legacy ciphers.
These calls have been updated to supply both the default and legacy providers.

##### Checklist

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: crypto/idea/build.info [hunk 1]
```diff
@@ -1,3 +1,10 @@
...
-        i_cbc.c i_cfb64.c i_ofb64.c i_ecb.c i_skey.c
+$ALL=i_cbc.c i_cfb64.c i_ofb64.c i_ecb.c i_skey.c
+
+SOURCE[../../libcrypto]=$ALL
+
+# When all deprecated symbols are removed, libcrypto doesn't export the
+# idea functions, so we must include them directly in liblegacy.a
+IF[{- $disabled{'deprecated-3.0'} -}]
+  SOURCE[../../providers/liblegacy.a]=$ALL
+ENDIF
```
Comment (levitte) [inline crypto/idea/build.info]: Nit: add a line ending

File: crypto/rc2/build.info [hunk 2]
```diff
@@ -1,3 +1,11 @@
...
+
+$ALL=rc2_ecb.c rc2_skey.c rc2_cbc.c rc2cfb64.c rc2ofb64.c
+
+SOURCE[../../libcrypto]=$ALL
+
+# When all deprecated symbols are removed, libcrypto doesn't export the
+# rc2 functions, so we must include them directly in liblegacy.a
+IF[{- $disabled{'deprecated-3.0'} -}]
+  SOURCE[../../providers/liblegacy.a]=$ALL
+ENDIF
```
Comment (levitte) [inline crypto/rc2/build.info]: Here as well

File: crypto/seed/build.info [hunk 3]
```diff
@@ -1,2 +1,10 @@
...
-SOURCE[../../libcrypto]=seed.c seed_ecb.c seed_cbc.c seed_cfb.c seed_ofb.c
+$ALL=seed.c seed_ecb.c seed_cbc.c seed_cfb.c seed_ofb.c
+
+SOURCE[../../libcrypto]=$ALL
+
+# When all deprecated symbols are removed, libcrypto doesn't export the
+# seed functions, so we must include them directly in liblegacy.a
+IF[{- $disabled{'deprecated-3.0'} -}]
+  SOURCE[../../providers/liblegacy.a]=$ALL
+ENDIF
```
Comment (levitte) [inline crypto/seed/build.info]: And here

**Comment by slontis** [comment]
The '''no-asm no-deprecated''' travis build had a few issues..

**Comment by slontis** [comment]
will look - I am curently fixing the mem ory leaks cause by using -provider in the apps.

**Comment by levitte** [comment]
Have a look at Travis, you have a memleak that's wreaking havoc...

**Comment by slontis** [comment]
Fixed pre existing memory leak (functionality had not been used yet in a test).

The reason the evp_tests all work is because it loads a default_and_legacy.cnf file. Manually testing a file confrms that it fails if the legacy provider is not loaded.
A seperate PR to clean up this stuff will need to be added. The availableIn is going to have to be reviewed for every test.
Basically it should have configs for
default_only.
fips_only.
legacy_only.
default+legacy
fips+legacy.

Then we would iterate over all of these..

Note some tests do for example RSA + MD5 .

**Comment by richsalz** [comment]
In regards to https://github.com/openssl/openssl/pull/11419#issuecomment-605428630, I was trying to do that (as discussed on the fips-sponsors mailing list).  See #11124.  Unfortunately, the pre-requisite PR's for that  -- #11369, #11347, #11177 etc -- are not being reviewed. :(  So I am moving on to something that will get review attention, "fetchable RAND"

While I'm here: this situation sucks, frankly.

**Comment by slontis** [comment]
ping

**Comment by paulidale** [comment]
re-ping @openssl/committers

**Comment by t-j-h** [comment]
I don't see any issues with the PR as it stands - but @levitte or @mattcaswell should be the ones reviewing this. If it remains unaddressed for much longer I will approve it myself.

File: test/recipes/20-test_enc.t [hunk 4]
```diff
@@ -26,6 +26,8 @@ my $testsrc = srctop_file("test","recipes",basename($0));
 my $test = catfile(".", "p");
 
 my $cmd = "openssl";
+my $bldtop = bldtop_dir("providers");
+my @prov = ("-provider_path", $bldtop, "-provider", "default", "-provider", "legacy");
```
Comment (levitte) [inline test/recipes/20-test_enc.t:30]: That's an odd name for a subdirectory of the build top. How about `$providers`?

File: test/recipes/20-test_enc_more.t [hunk 5]
```diff
@@ -25,6 +25,8 @@ my $cipherlist = undef;
 my $plaintext = catfile(".", "testdatafile");
 my $fail = "";
 my $cmd = "openssl";
+my $bldtop = bldtop_dir("providers");
+my @prov = ("-provider_path", $bldtop, "-provider", "default", "-provider", "legacy");
```
Comment (levitte) [inline test/recipes/20-test_enc_more.t:29]: Same here

File: test/recipes/80-test_cms.t [hunk 6]
```diff
@@ -13,14 +13,17 @@ use warnings;
...
+use OpenSSL::Test qw/:DEFAULT srctop_dir srctop_file bldtop_dir/;
 use OpenSSL::Test::Utils;
 
 setup("test_cms");
 
 plan skip_all => "CMS is not supported by this OpenSSL build"
     if disabled("cms");
 
+my $bldtop = bldtop_dir("providers");
+my @prov = ("-provider_path", $bldtop, "-provider", "default", "-provider", "legacy");
```
Comment (levitte) [inline test/recipes/80-test_cms.t:25]: And here...

File: test/recipes/20-test_enc.t (hunk 4 above)
Comment (slontis) [reply test/recipes/20-test_enc.t:30]: Changed all 3 files to use $provpath

**Comment by levitte** [comment]
Gah!  All those timeouts (yeahok, only two)

**Comment by bernd-edlinger** [comment]
I'd prefer a penguin over a bitten apple.  :smile:

**Comment by slontis** [comment]
Thanks. Merged to master.

**Comment by t8m** [comment]
@slontis This (most probably) caused a regression in one of the Travis builds which was added when PR #11468 was merged.
https://travis-ci.org/github/openssl/openssl/jobs/672855572

**Comment by t8m** [comment]
The reason it was not caught in this PR CI run is that there was no no-deprecated build without no-asm.

**Pull request closed without merging**